File: clj/src/chapter_1/discount.clj
(ns chapter_1.discount)

(def prices [10 30 17 20 15 18 45 12])

(defn discount [over dscnt]
  (reduce + 0 (map #(* dscnt %) (filter #(> % over) prices))))

(defn discountTL [over dscnt]
  (->> prices
       (filter #(> % over))
       (map #(* dscnt %))
       (reduce + 0)))

File: scala/src/chapter_1/Discount.scala
package chapter_1

object Discount {
  val prices: List[BigDecimal] = List(10, 30, 17, 20, 15, 18, 45, 12)

  def discount (over: BigDecimal, percent: BigDecimal) = {
    prices
      .filter(_ > over)
      .map(_ * percent)
      .foldLeft(BigDecimal(0))(_ + _)
  }
}